### PR TITLE
Rolling ROS image is now based on "noble"

### PR DESCRIPTION
**Pull request by amarburg**
## Changes Made

I am not tracking the underlying changes to the ROS image in great detail, @evan-palmer, you may have more info or a better perspective on how to resolve this issue.

However, it looks like Docker images for "rolling" are now built on Ubuntu "noble."

This brings a few major changes, some of which were also addressed when [porting to "jazzy"](https://github.com/apl-ocean-engineering/blue/pull/1)

This PR pulls out the elements specific to "noble" without the jazzy-specific adaptations (particularly building MAVROS from source).

Specifically:

* Use the existing non-root user "ubuntu" (uid/gid 1000) which already exists in the image.
* Python packages installed with `pip` are placed in a venv

**Please note:** this image must use Gazebo "harmonic," so it is based on #223 .   Complete and merge that PR before merging this PR.

## Testing

TBD

### CODE REVIEW

**Comment by evan-palmer** [comment]
Thanks for submitting this @amarburg! Have you run into any MAVROS/Noble compatibility issues?

**Comment by amarburg** [comment]
I’m still testing … both MAVROS and Gazebo.  TBD…

**Comment by amarburg** [comment]
Split the "use Gazebo Harmonic" parts into #223.

**Comment by amarburg** [comment]
Added note that this PR is based on #223 .   Complete that PR first.

**Comment by amarburg** [comment]
Update:  testing #223 in Iron, I am able to do the keyboard and joystick demos in simulation.

In this PR (which includes #223) I get no response in simulation in Jazzy or Rolling.   So... time to debug.

**Comment by amarburg** [comment]
> In this PR (which includes #223) I get no response in simulation in Jazzy or Rolling. So... time to debug.

Follow up.  This is all in an `-desktop-nvidia` image, all in Harmonic.   Iron images are built in #223 , while rolling images are built in this PR (#220 )

It appears the `ardusub` is not publishing `LOCAL_POSITION_NED` (msgid: 32), so the `/mavros/local_position/velocity` message is not published and the ISMC never runs.

By comparison, under Iron, IMU initialization looks like:

```
[mavros_node-2] [INFO] [1723059285.162502726] [mavros.sys]: FCU: EKF3 IMU0 initialised
[mavros_node-2] [INFO] [1723059285.162589783] [mavros.sys]: FCU: EKF3 IMU1 initialised
[mavros_node-2] [INFO] [1723059285.168403259] [mavros.sys]: FCU: AHRS: EKF3 active
[mavros_node-2] [INFO] [1723059287.667229027] [mavros.sys]: FCU: EKF3 IMU1 tilt alignment complete
[mavros_node-2] [INFO] [1723059287.667314698] [mavros.sys]: FCU: EKF3 IMU1 is using external nav data
[mavros_node-2] [INFO] [1723059287.667367775] [mavros.sys]: FCU: EKF3 IMU1 initial pos NED = 1.3,-0.1,0.5 (m)
[mavros_node-2] [INFO] [1723059287.670126325] [mavros.sys]: FCU: EKF3 IMU0 tilt alignment complete
[mavros_node-2] [INFO] [1723059287.670213435] [mavros.sys]: FCU: EKF3 IMU0 is using external nav data
[mavros_node-2] [INFO] [1723059287.670285123] [mavros.sys]: FCU: EKF3 IMU0 initial pos NED = 1.3,-0.1,0.5 (m)
[mavros_node-2] [INFO] [1723059287.699801308] [mavros.sys]: FCU: EKF3 IMU1 yaw aligned
[mavros_node-2] [INFO] [1723059287.701801746] [mavros.sys]: FCU: EKF3 IMU0 yaw aligned
```

with Rolling:

```
mavros_node-2] [INFO] [1723061971.085722987] [mavros.sys]: FCU: EKF3 IMU0 initialised
[mavros_node-2] [INFO] [1723061971.085800486] [mavros.sys]: FCU: EKF3 IMU1 initialised
[mavros_node-2] [INFO] [1723061971.087894765] [mavros.sys]: FCU: AHRS: EKF3 active
[mavros_node-2] [INFO] [1723061973.462258874] [mavros.sys]: FCU: EKF3 IMU1 tilt alignment complete
[mavros_node-2] [INFO] [1723061973.465137499] [mavros.sys]: FCU: EKF3 IMU0 tilt alignment complete
```

which is missing the "using external nav data" message.

**Comment by amarburg** [comment]
I believe the mavros yaml-cpp linkage issue mandates building mavros from source on _both_ rolling and jazzy.   In that case, this PR subsumes #169 , there's no difference.  

Instead just backport this PR to the jazzy branch.

TBD still testing.

**Comment by amarburg** [comment]
OK, I think we're there.   To recap, this PR updates "rolling" and "jazzy" builds, which are both based on Ubuntu "noble."   It:

* Uses the user "ubuntu" (UID 1000) provided by the base image rather than creating the  new user "blue"
   * Updates dev containers for this new username
* Creates a venv "blue" for packages pip-installed by the user.  This venv is activated automatically through a line in `.bashrc`
* Builds mavros and mavlink from source -- there are no packages for jazzy and library link issues with the packages for rolling.   Includes two patches to the mavros source tree to fix python "future" and aforementioned link issues.
* Configures the ardupilot `install-prereqs-ubuntu.sh ` script to _not_ install STM32 crosscompile tools and to _not_ activate the ardupilot venv by default (presumably they create this venv in newer Ubuntu images for the same reasons as above)

File: .docker/Dockerfile
Comment (evan-palmer) [inline .docker/Dockerfile:121]: Applying a patch seems fine, but adds some extra files to maintain (and adds more required changes when we eventually install the binary). Here is the sed command for this patch:
```bash
sed -i -e 's/^from future import standard_library/#from future import standard_library/' \
          -e 's/standard_library.install_aliases()/#standard_library.install_aliases()/' \
          <file-name>
```
Not sure what the filename is here, so that would need to be fixed.

File: .docker/Dockerfile
Comment (evan-palmer) [inline .docker/Dockerfile:123]: sed command for this patch:
```bash
sed -i -e 's/^# find_package(yaml_cpp REQUIRED)/find_package(yaml-cpp REQUIRED)/' \
       -e '/^ament_target_dependencies(mavros_extras_plugins/i target_link_libraries(mavros_extras_plugins yaml-cpp::yaml-cpp)' \
       -e '/^ament_target_dependencies(mavros_extras/i target_link_libraries(mavros_extras yaml-cpp::yaml-cpp)' \
       <file-name>
```

File: .docker/patches/mavlink_standard_library.patch
Comment (evan-palmer) [inline .docker/patches/mavlink_standard_library.patch:1]: We can remove this in favor of the sed command

File: .docker/patches/mavros_extra_yaml_linkage.patch
Comment (evan-palmer) [inline .docker/patches/mavros_extra_yaml_linkage.patch:1]: ditto

File: .dockerignore
Comment (evan-palmer) [inline .dockerignore:1]: ditto

**Review by evan-palmer — CHANGES_REQUESTED** [review]
Looks good overall! I attached the sed commands to apply those patches. Lets add an issue when this is merged to track the Jazzy/Rolling binaries for MAVROS.

File: .dockerignore
Comment (amarburg) [reply .dockerignore:1]: Removed

File: .docker/patches/mavros_extra_yaml_linkage.patch
Comment (amarburg) [reply .docker/patches/mavros_extra_yaml_linkage.patch:1]: Removed

File: .docker/patches/mavlink_standard_library.patch
Comment (amarburg) [reply .docker/patches/mavlink_standard_library.patch:1]: Removed

File: .docker/Dockerfile
Comment (amarburg) [reply .docker/Dockerfile]: Excellent.  Thanks!

File: .docker/Dockerfile
Comment (amarburg) [reply .docker/Dockerfile]: Done!

**Comment by amarburg** [comment]
> Looks good overall! I attached the sed commands to apply those patches. Lets add an issue when this is merged to track the Jazzy/Rolling binaries for MAVROS.

Thanks for writing those.   Replaced the patches with sed....

**Review by evan-palmer — APPROVED** [review]
Just waiting on the CI to finish then this should be good to merge and backported to Jazzy

**Comment by evan-palmer** [comment]
Hmm. It looks like the `robot` arm64 build is failing due to a timeout (it's taking about 6 hours). This has happened across the past few builds. The main contributor seems to be the MAVROS build, which takes about 4.5 hours before it is canceled. I'll look to see if we could try doing some cache optimization, but I don't know that there is much we can do here.

**Comment by amarburg** [comment]
> Hmm. It looks like the `robot` arm64 build is failing due to a timeout (it's taking about 6 hours). This has happened across the past few builds. The main contributor seems to be the MAVROS build, which takes about 4.5 hours before it is canceled. I'll look to see if we could try doing some cache optimization, but I don't know that there is much we can do here.


Hm, yeah.   The arm64 cross-compile is bruuuutal.   Just for reference let me time a build on my desktop.

**Comment by evan-palmer** [comment]
It looks like we can use the [GitHub Actions cache](https://docs.docker.com/build/cache/backends/gha/) to help with this. I see that you're also testing out Buildx bake @amarburg

**Comment by amarburg** [comment]
> It looks like we can use the [GitHub Actions cache](https://docs.docker.com/build/cache/backends/gha/) to help with this. I see that you're also testing out Buildx bake @amarburg

Yes, using it for my local builds.   Still testing the Github actions integration.

**Comment by amarburg** [comment]
> Hm, yeah. The arm64 cross-compile is bruuuutal. Just for reference let me time a build on my desktop.

```
docker buildx bake  65.32s user 37.25s system 0% cpu 8:17:02.73 total
```

**Comment by evan-palmer** [comment]
> 8:17:02.73 total

:open_mouth:

**Comment by amarburg** [comment]
I'm working on putting mavros in a build stage.  It makes this PR a lot less surgical and starts to mush into #226.

Is it worth letting this PR run to completion once, and merge it (even if it means extending the timeout on CI)?  Then work on mavros build caching/acceleration in a separate PR.

**Comment by evan-palmer** [comment]
I think that's fine. To run this to completion, we will need to extend the default job timeout for the arm64 build. By default it is set to 6 hours and 30 min. I've added a review comment with the relevant configuration.

**Comment by evan-palmer** [comment]
[This](https://github.com/orgs/community/discussions/26679) thread seems to indicate that we have to set the timeout at the job level rather than the step level for the timeout to apply

**Comment by amarburg** [comment]
Hmmm...

> #49 15081.8 Finished <<< mavros_msgs [4h 10min 54s]

4h for a `*_msgs` repo?   And it never actually finishes mavros.

**Comment by evan-palmer** [comment]
I'm surprised that `mavros_msgs` is the main bottleneck here. I would say that we could try some different CMake flags, but I don't know that we will make much of a dent there. 

The [Jenkins build](https://build.ros2.org/job/Idev__mavros__ubuntu_jammy_amd64/25/) completes in 47 minutes, which I believe is a multi-architecture build. Maybe there's something we're missing here.

Another alternative could be to build MAVROS in a fork, then store the build as a GitHub package. That seems like unnecessary effort though. I'll ping some folks to see what other options we have.

**Comment by evan-palmer** [comment]
> Is it worth letting this PR run to completion once, and merge it (even if it means extending the timeout on CI or only building for amd64)? Then work on mavros build caching/acceleration in a separate PR.

I think we can go ahead and merge this and focus on addressing the `robot` build in #241

File: .github/workflows/docker.yaml
Comment (evan-palmer) [inline .github/workflows/docker.yaml:68]: We can remove this, then merge

**Review by evan-palmer — CHANGES_REQUESTED** [review]

**Review by evan-palmer — COMMENTED** [review]
We'll need to add the following line to the `.vscode/settings.json` file too:
```json
  "python.defaultInterpreterPath": "${workspaceFolder}/.venv/online-dynamics-learning/bin/python",
```

**Comment by amarburg** [comment]
* Disabled `arm64` build in the Github action
* Removed the timout
* Added the `python.defaultInterpreterPath` to `.vscode/settings.json`.

**Comment by evan-palmer** [comment]
Thanks for all of your work on this! I'll get it merged and backport to Jazzy

**Review by evan-palmer — APPROVED** [review]

**Comment by evan-palmer** [comment]
@mergify backport jazzy

**Pull request merged**
